New build command

tsandall opened the pull request:
This PR updates the `opa build` command to address #1460 and improve support for packaging policy and data files into bundles. This PR doesn't have any docs because the basic interaction is described inside the help text for the command. I think we should review the existing docs and see how `opa build` could be incorporated, but that can come in a subsequent PR. Here's the help text:

```
Build an OPA bundle.

The 'build' command packages OPA policy and data files into bundles. Bundles are
gzipped tarballs containing policies and data. Paths referring to directories are
loaded recursively.

	$ ls
	example.rego

	$ opa build -b .

You can load bundles into OPA on the command-line:

	$ ls
	bundle.tar.gz example.rego

	$ opa run bundle.tar.gz

You can also configure OPA to download bundles from remote HTTP endpoints:

	$ opa run --server \
		--set bundles.example.resource=bundle.tar.gz \
		--set services.example.url=http://localhost:8080

Inside another terminal in the same directory, serve the bundle via HTTP:

	$ python3 -m http.server --bind localhost 8080

For more information on bundles see https://www.openpolicyagent.org/docs/latest/management.

When -b is specified the 'build' command assumes paths refer to existing bundle files
or directories following the bundle structure. If multiple bundles are provided, their
contents are merged. If there are any merge conflicts (e.g., due to conflicting bundle
roots), the command fails.

The -O flag controls the optimization level. By default, optimization is disabled (-O=0).
When optimization is enabled the 'build' command generates a bundle that is semantically
equivalent to the input files however the structure of the files in the bundle may have
been changed by rewriting, inlining, pruning, etc. Higher optimization levels may result
in longer build times.

The 'build' command supports targets (specified by -t):

    rego    The default target emits a bundle containing a set of policy and data files
            that are semantically equivalent to the input files. If optimizations are
            disabled the output may simply contain a copy of the input policy and data
            files. If optimization is enabled at least one entrypoint (-e) must be supplied.

    wasm    The wasm target emits a bundle containing a WebAssembly module compiled from
            the input files. The bundle may contain the original policy or data files.
            The wasm target requires exactly one entrypoint (-e) be supplied.

The -e flag tells the 'build' command which documents will be queried by the software
asking for policy decisions, so that it can focus optimization efforts and ensure
that document is not eliminated by the optimizer.

```

timothyhinrichs commented on cmd/build.go:
```diff
@@ -5,94 +5,172 @@
+            the input files. The bundle may contain the original policy or data files.
+            The wasm target requires exactly one entrypoint (-e) be supplied.
+
+The -e flag tells the 'build' command that a specific document may be queried by the
```
I read this line a couple times before understanding it.  Maybe....

The -e flag tells the 'build' command which documents will be queried by the software asking for policy decisions, so that it can focus optimization efforts and ensure that document is not eliminated by the optimizer.

tsandall replied:
Fixed.

tsandall commented on bundle/bundle.go:
```diff
@@ -62,6 +62,49 @@ func (m *Manifest) Init() {
+func (m Manifest) Equal(other Manifest) bool {
+
+	// This is safe since both are passed by value.
+	m.Init()
+	other.Init()
```
Yeah, I think goal was to make the zero value valid--not sure whether it's useful but that's how we've tried to treat it elsewhere AFAIK. Trying to avoid changing that here.

tsandall commented on bundle/bundle.go:
```diff
@@ -454,6 +532,53 @@ func mktree(path []string, value interface{}) (map[string]interface{}, error) {
 }
 
+// Merge accepts a set of bundles and merges them into a single result bundle. If there are
+// any conflicts during the merge (e.g., with roots) an error is returned. The result bundle
+// will have an empty revision except in the special case where a single bundle is provided.
+func Merge(bundles []*Bundle) (*Bundle, error) {
```
Hrmmm. I don't have a strong opinion. I'd probably just leave it as is, unless there's some reason why you think it's bad--I could see the argument that it's simpler to just erase the revision in all cases--then callers _have_ to deal with it...I'm just not sure it simplifies things for the caller...with this approach they can still reset the revision in all cases if they want.

patrick-east replied:
> I could see the argument that it's simpler to just erase the revision in all cases--then callers have to deal with it...I'm just not sure it simplifies things for the caller

Well, I guess that simple != least amount of code for the caller. I was thinking that making the caller always have to deal with it seems more likely to result in correct revisions versus them needing to have read the docs and/or understand what it is doing.

tsandall replied:
I've updated the godoc to indicate that in the special case, the bundle is returned unmodified. Thinking about it more, I prefer this because otherwise we need to define properties that determine if a single bundle should be modified at all. For example, should the roots be rewritten? If we add other fields in the future, we could have similar questions to answer. Leaving the bundle unmodified feels like the simplest way to define a "merge" operation on a single bundle. WDYT?

patrick-east replied:
Works for me 👍

tsandall commented on rego/rego.go:
```diff
@@ -1776,7 +1783,8 @@ func (r *Rego) partial(ctx context.Context, ectx *EvalContext) (*PartialQueries,
 		WithIndexing(ectx.indexing).
-		WithPartialNamespace(ectx.partialNamespace)
+		WithPartialNamespace(ectx.partialNamespace).
+		WithSkipPartialNamespace(r.skipPartialNamespace)
```
Added test.

tsandall commented on cmd/build.go:
```diff
@@ -5,94 +5,173 @@
-		for _, bundleDir := range buildParams.bundlePaths.v {
-			regoArgs = append(regoArgs, rego.LoadBundle(bundleDir))
-		}
+	buildCommand.Flags().VarP(buildParams.target, "target", "t", "set the output bundle target type")
+	buildCommand.Flags().BoolVarP(&buildParams.debug, "debug", "", false, "enable debug output")
+	buildCommand.Flags().IntVarP(&buildParams.optimizationLevel, "optimize", "O", 0, "set optimization level")
+	buildCommand.Flags().VarP(&buildParams.entrypoints, "entrypoint", "e", "set slash separated entrypoint path")
+	buildCommand.Flags().StringVarP(&buildParams.revision, "revision", "r", "", "set output bundle revision")
+	buildCommand.Flags().StringVarP(&buildParams.outputFile, "output", "o", "bundle.tar.gz", "set the output filename")
+	addBundleModeFlag(buildCommand.Flags(), &buildParams.bundleMode, false)
+	addIgnoreFlag(buildCommand.Flags(), &buildParams.ignore)
+	RootCommand.AddCommand(buildCommand)
```
Fixed docs to show new build command.

tsandall commented on bundle/bundle.go:
```diff
@@ -62,6 +62,13 @@ func (m *Manifest) Init() {
 }
 
+// AssertRoot ensures the root r is contained in the manifest, creating it if not.
+func (m *Manifest) AssertRoot(r string) {
```
Renamed to AddRoot and added test case.

tsandall commented on compile/compile_test.go:
```diff
@@ -0,0 +1,834 @@
+	test.WithTempFS(nil, func(root string) {
+		err := New().WithPaths(path.Join(root, "does-not-exist")).Build(context.Background())
+		if err == nil {
+			t.Fatal(err)
```
Fixed error message.

tsandall commented on compile/compile.go:
```diff
@@ -0,0 +1,603 @@
+	// Level 2: L1 except inlining of rules with unknowns.
+	// Level 3: L2 except aggressive inlining using negation and copy propagation optimizations.
+
+	// NOTE(tsandall): if there are multiple entrypoints, copy the bundle because
+	// if any of the optimization steps fail, we do not want to leave the caller's
+	// bundle in a partially modified state.
```
I don't have a strong opinion here. Since the caller is internal only for now, I'd be inclined to just leave it and avoid the unnecessary deep copy of all of the modules and all of the data.

tsandall commented on compile/compile.go:
```diff
@@ -0,0 +1,603 @@
+
+func (c *Compiler) compileWasm(ctx context.Context) error {
+
+	if c.compiler == nil {
+		var err error
+		c.compiler, err = compile(c.bundle)
+		if err != nil {
+			return err
+		}
+	}
```
Added a comment to explain why we lazily initialize the compiler.

tsandall commented on compile/compile.go:
```diff
@@ -0,0 +1,603 @@
+	}
+
+	// NOTE(tsandall): do not use the compiler from the optimizer as it will
+	// be stale and not reflect the output bundle contents.
```
Fixed to have the optimizer just expose an API that lets the caller retrieve the bundle.

tsandall commented on compile/compile.go:
```diff
@@ -0,0 +1,603 @@
+// WithOptimizationLevel sets the optimization level on the builder. By default
+// optimizations are disabled. Higher levels apply more aggressive optimizations
+// but take longer.
+func (c *Compiler) WithOptimizationLevel(n int) *Compiler {
```
Updated comment.

tsandall commented on compile/compile.go:
```diff
@@ -0,0 +1,603 @@
+	return fmt.Sprintf("%v: %v", d.Location, d.Message)
+}
+
+// New returns a new builder instance that can be invoked.
```
Fixed.

tsandall commented on bundle/bundle.go:
```diff
@@ -477,6 +602,8 @@ func RootPathsOverlap(pathA string, pathB string) bool {
 func insertValue(b *Bundle, path string, value interface{}) error {
 
+	// NOTE(tsandall)
+
```
Not sure what happened here, lol. It probably wasn't important. Didn't want to waste time on the delete.

tsandall commented on bundle/bundle.go:
```diff
@@ -454,6 +532,53 @@ func mktree(path []string, value interface{}) (map[string]interface{}, error) {
+
+		for _, root := range *b.Manifest.Roots {
+			key := strings.Split(root, "/")
+			if val := b.read(key); val != nil {
```
Renamed to readDat() and insertData().

tsandall commented on bundle/bundle.go:
```diff
@@ -454,6 +532,53 @@ func mktree(path []string, value interface{}) (map[string]interface{}, error) {
+			// NOTE(tsandall): This could be relaxed in the future to allow, e.g., combination
+			// of wasm bundles with others that only contain base documents.
+			return nil, errors.New("wasm bundles cannot be merged")
+		}
```
Moved into godoc.

tsandall commented on bundle/bundle.go:
```diff
@@ -454,6 +532,53 @@ func mktree(path []string, value interface{}) (map[string]interface{}, error) {
+
+	for _, b := range bundles {
+		// NOTE(tsandall): If the roots are non-nil this indicates an error in the caller.
+		roots = append(roots, *b.Manifest.Roots...)
```
Fixed. Comment was misleading and we could just return an error like you said--added a test case.

tsandall commented on bundle/bundle.go:
```diff
@@ -62,6 +62,49 @@ func (m *Manifest) Init() {
 	}
 }
 
+// Equal returns true if is semantically equivalent to other.
```
Fixed.

tsandall commented on bundle/bundle.go:
```diff
@@ -304,7 +329,11 @@ func Write(w io.Writer, bundle Bundle) error {
+		if w.usePath {
+			path = module.Path
+		}
+		if err := archive.WriteFile(tw, path, module.Raw); err != nil {
```
Yep, good idea.

tsandall commented on bundle/bundle_test.go:
```diff
@@ -265,6 +266,7 @@ func TestRoundtrip(t *testing.T) {
 		},
 		Modules: []ModuleFile{
 			{
+				URL:    "/foo/corge/corge.rego",
```
This is handled by the test for usePath as false.

tsandall commented:
> We probably should add a comment for the changelog about the breaking changes to opa build, not sure if it needs to be with this PR.. but we want to make sure it doesn't slip through the cracks come release time.

yes, we should definitely include a note. The changelog script will report all of these commits so I don't expect it to go unnoticed.

> Part of me wonders if it's worth deprecating the old opa build, maybe even hide it, and making this new stuff opa compile or something so we don't break anyone who is using the WASM compilation (or the blogs/presentations/readme/etc that showed how to use it). It's probably not a lot of users, but I can imagine that as people start to pick it up it will be bad to have outdated instructions like that floating around.

I'd like to go ahead w/ the breaking change because I feel like `opa build` is too good to give up. I think we can rollout the changes to other repos fairly easily. If anything this might accelerate some development in the node SDK to implement bundle downloading 😅

patrick-east commented:
> potential breakages include: node SDK, contrib repo (for the cloudflare example)...what else?

I think thats it for projects we own

tsandall commented:
I've squashed the fixes into their corresponding commits. Also, I've squashed the bundle package commits into one because they were dependent on each other.

tsandall commented on bundle/bundle_test.go:
```diff
@@ -393,3 +586,208 @@ func TestParsedModules(t *testing.T) {
+					"baz": "val2",
+				},
+			},
+		},
```
Yes, we have coverage of this case...though it's a bit subtle. The merge test ensures that conflicting roots are not allowed and the bundle/manifest validation itself ensures that the bundle does not contain data outside of the roots, so we should be guaranteed that conflicts are caught.

tsandall commented on bundle/bundle.go:
```diff
@@ -286,9 +350,45 @@ func (r *Reader) fullPath(path string) string {
 }
 
-// Write serializes the Bundle and writes it to w.
+// Write is deprecated. Use Writer instead.
```
Fixed.

tsandall commented on bundle/bundle.go:
```diff
@@ -286,9 +350,45 @@ func (r *Reader) fullPath(path string) string {
+}
+
+// UseModulePath configures the writer to use the module file path instead of the
+// module file URL during serialization. This is for bacwkards compatibility.
```
Fixed.

DTrejo commented:
TSandall, thanks for your work on opa.

We were using the old version of the build command  to compile a tree directories with rego files and get the output of all of them as a JSON object. 

How can I accomplish this with the new build command? I wouldn't ask except that I updated the `opa` command on homebrew and there's no old version in brew to downgrade to. 

Thanks in advance,
D

The rabbit hole:
Why did I upgrade in the first place? count() did not work in WASM.  I saw you fixed that.
Why did I need count? I wanted to negate in a short way using a set: 
```rego
any_false = {req | req := required[_]; req.allow == false}
count(any_false) == 0
```
Why did I negate using a set? Because `not (required[_].allow == false)` doesn't work as a one-liner. I upvoted #2436 .
😆

tsandall commented:
@DTrejo can you share the exact `opa build` command you were running before?

If it was something like `opa build -d <directory-path> 'data.foo.bar=x' then you could just say:

```
opa build <directory-path> -t wasm -e foo/bar
```

DTrejo commented:
Thank you for the quick reply  :) I am using it like this:

`opa build -d . -o all.wasm data=results`

Directory structure:
```
a-directory/
   1-file.rego
  2-file.rego
b-directory/
  1-file.rego
  2-file.rego
all.wasm
```

tsandall commented:
@DTrejo I see. We don't normally see people querying for _all of_ `data` because it's often desirable to just ask for the value of one document (which could be defined by a rule and aggregate as much information as needed.) Because of that, the --entrypoint flag expects the path to be non-empty--however, I don't see a reason why we could not relax that. I've filed an issue to track that enhancement.

One option right now would be to define a rule that aggregates the values from the rules that you have. For example, if the files you have contribute to packages `a.b.c`, `a.b.d`, and `b.e`, you could write a rule like:

```rego
package main

import data.a.b.c
import data.a.b.d
import data.b.e

result = {"a": {"b": {"c": c, "d": d}}, "b": {"e": e}}
```

Or even more simply:

```rego
package main

result = {"a": data.a, "b": data.b}
```

Then when you run `opa build` you could say:

```
opa build . -t wasm -e main/result
```

DTrejo commented:
That is a good solution, thank you! I ended up going back to .19 for now, but that will be useful to know in the future for my team! Have a great week!
